src/colmap/controllers/feature_matching.cc

```diff
@@ -47,523 +46,75 @@ void PrintElapsedTime(const Timer& timer) {
-  struct Retrieval {
-    image_t image_id = kInvalidImageId;
-    std::vector<retrieval::ImageScore> image_scores;
-  };
-
-  // Create a thread pool to retrieve the nearest neighbors.
-  ThreadPool retrieval_thread_pool(num_threads);
-  JobQueue<Retrieval> retrieval_queue(num_threads);
-
-  // The retrieval thread kernel function. Note that the descriptors should be
-  // extracted outside of this function sequentially to avoid any concurrent
-  // access to the database causing race conditions.
-  retrieval::VisualIndex<>::QueryOptions query_options;
-  query_options.max_num_images = num_images;
-  query_options.num_neighbors = num_neighbors;
-  query_options.num_checks = num_checks;
-  query_options.num_images_after_verification = num_images_after_verification;
-  auto QueryFunc = [&](const image_t image_id) {
-    auto keypoints = *cache->GetKeypoints(image_id);
-    auto descriptors = *cache->GetDescriptors(image_id);
-    if (max_num_features > 0 && descriptors.rows() > max_num_features) {
-      ExtractTopScaleFeatures(&keypoints, &descriptors, max_num_features);
-    }
-
-    Retrieval retrieval;
-    retrieval.image_id = image_id;
-    visual_index->Query(
-        query_options, keypoints, descriptors, &retrieval.image_scores);
-
-    THROW_CHECK(retrieval_queue.Push(std::move(retrieval)));
-  };
-
-  // Initially, make all retrieval threads busy and continue with the matching.
-  size_t image_idx = 0;
-  const size_t init_num_tasks =
-      std::min(image_ids.size(), 2 * retrieval_thread_pool.NumThreads());
-  for (; image_idx < init_num_tasks; ++image_idx) {
-    retrieval_thread_pool.AddTask(QueryFunc, image_ids[image_idx]);
-  }
-
-  std::vector<std::pair<image_t, image_t>> image_pairs;
-
-  // Pop the finished retrieval results and enqueue them for feature matching.
-  for (size_t i = 0; i < image_ids.size(); ++i) {
-    if (thread->IsStopped()) {
-      retrieval_queue.Stop();
-      return;
-    }
-
-    Timer timer;
-    timer.Start();
-
-    LOG(INFO) << StringPrintf("Matching image [%d/%d]", i + 1, image_ids.size())
-              << std::flush;
-
-    // Push the next image to the retrieval queue.
-    if (image_idx < image_ids.size()) {
-      retrieval_thread_pool.AddTask(QueryFunc, image_ids[image_idx]);
-      image_idx += 1;
-    }
-
-    // Pop the next results from the retrieval queue.
-    auto retrieval = retrieval_queue.Pop();
-    THROW_CHECK(retrieval.IsValid());
-
-    const auto& image_id = retrieval.Data().image_id;
-    const auto& image_scores = retrieval.Data().image_scores;
-
-    // Compose the image pairs from the scores.
-    image_pairs.clear();
-    image_pairs.reserve(image_scores.size());
-    for (const auto image_score : image_scores) {
-      image_pairs.emplace_back(image_id, image_score.image_id);
-    }
-
-    matcher->Match(image_pairs);
-
-    PrintElapsedTime(timer);
-  }
-}
-
-class ExhaustiveFeatureMatcher : public Thread {
+template <typename DerivedPairGenerator>
```

Comment:
I guess instead of templating we could pass an abstract pair generator and the cache object from the outside. Not necessarily important for here but theoretically has some benefits for injecting a mock/fake for testing purposes of this class.